src/Tasks/Microsoft.NET.Build.Tasks/GetPackagesToPrune.cs

```diff
@@ -72,51 +83,208 @@ protected override void ExecuteCore()
+                        packagesForFrameworkReference = LoadPackagesToPruneFromFrameworkPackages(key.TargetFrameworkIdentifier, key.TargetFrameworkVersion, frameworkReference,
+                            acceptNearestMatch: true);
+                    }
+                }
+
+                if (packagesForFrameworkReference == null)
+                {
+                    //  We didn't find the data for packages to prune.  This indicates that there's a bug in the SDK construction, so fail hard here so that we fix that
+                    //  (rather than a warning that might be missed).
+                    //  Since this indicates an error in the SDK build, the message probably doesn't need to be localized.
+
+                    if (allowMissingPrunePackageData)
+                    {
+                        log.LogMessage($"Prune package data not found for {key.TargetFrameworkIdentifier} {key.TargetFrameworkVersion} {frameworkReference}");
+                    }
+                    else
+                    {
+                        log.LogError(Strings.PrunePackageDataNotFound, key.TargetFrameworkIdentifier, key.TargetFrameworkVersion, frameworkReference);
+                    }
+                }
+                else
+                {
+                    AddPackagesToPrune(packagesToPrune, packagesForFrameworkReference.Select(kvp => (kvp.Key, kvp.Value)), log);
                 }
             }
 
-            PackagesToPrune = packagesToPrune.Select(p =>
+            return packagesToPrune.Select(p =>
             {
                 var item = new TaskItem(p.Key);
-                item.SetMetadata("Version", p.Value.ToString());
+
+                //  If a given version of a package is included in a framework, assume that any patches
+                //  to that package will be included in patches to the framework, and thus should be pruned.
+                //  See https://github.com/dotnet/sdk/issues/44566
+                //  To do this, we set the patch version for the package to be pruned to 32767, which should be
+                //  higher than any actual patch version.
+                var maxPatch = new NuGetVersion(p.Value.Major, p.Value.Minor, 32767);
+
+                item.SetMetadata("Version", maxPatch.ToString());
                 return item;
             }).ToArray();
+        }
 
-            BuildEngine4.RegisterTaskObject(key, PackagesToPrune, RegisteredTaskObjectLifetime.Build, true);
+        static Dictionary<string, NuGetVersion> LoadPackagesToPruneFromFrameworkPackages(string targetFrameworkIdentifier, string targetFrameworkVersion, string frameworkReference, bool acceptNearestMatch = false)
+        {
+            var nugetFramework = new NuGetFramework(targetFrameworkIdentifier, Version.Parse(targetFrameworkVersion));
+
+            //  FrameworkPackages just has data for .NET Framework 4.6.1 and doesn't handle framework compatibility, so treat anything greater than .NET Framework as if it were 4.6.1
```

Comment:
I think this does handle compatibility - https://github.com/dotnet/sdk/blob/b6fef1d9dc68e04ff93587328dfd45a707874943/src/Tasks/Microsoft.NET.Build.Tasks/FrameworkPackages/FrameworkPackages.cs#L97

Reply:
I disabled that in most cases because I want it to error out if it can't find a framework except in specific cases.  But I've changed this logic to re-enable the compatibility for .NET Framework instead of resetting the TargetFramework it's looking up.